src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -210,6 +210,110 @@ _server_monitor_append_cluster_time (mongoc_server_monitor_t *server_monitor,
    }
    mc_tpld_drop_ref (&td);
 }
+static bool
+_server_monitor_send_and_recv_hello_opmsg (
+   mongoc_server_monitor_t *server_monitor,
+   bson_t *cmd,
+   bson_t *reply,
+   bson_error_t *error)
+{
+   mongoc_rpc_t rpc = {0};
+   mongoc_array_t array_to_write;
+   mongoc_iovec_t *iovec;
+   int niovec;
+
+   mongoc_buffer_t buffer;
+   uint32_t reply_len;
+   bson_t temp_reply;
+   bool ret = false;
+
+   /* First, let's construct and send our OPCODE_MSG: */
+   rpc.header.msg_len = 0;
+   rpc.header.request_id = server_monitor->request_id++;
+   rpc.header.response_to = 0;
+   rpc.header.opcode = MONGOC_OPCODE_MSG;
+   rpc.msg.flags = 0;
+   rpc.msg.n_sections = 1;
+   rpc.msg.sections[0].payload_type = 0;
+   rpc.msg.sections[0].payload.bson_document = bson_get_data (cmd);
+
+   _mongoc_array_init (&array_to_write, sizeof (mongoc_iovec_t));
+   _mongoc_rpc_gather (&rpc, &array_to_write);
+
+   iovec = (mongoc_iovec_t *) array_to_write.data;
+   niovec = array_to_write.len;
+   _mongoc_rpc_swab_to_le (&rpc);
+
+   MONITOR_LOG (server_monitor,
+                "sending with timeout %" PRId64,
+                server_monitor->connect_timeout_ms);
+
+   if (!_mongoc_stream_writev_full (server_monitor->stream,
+                                    iovec,
+                                    niovec,
+                                    server_monitor->connect_timeout_ms,
+                                    error)) {
+      MONITOR_LOG_ERROR (
+         server_monitor, "failed to write awaitable hello: %s", error->message);
+      _mongoc_array_destroy (&array_to_write);
+      return false;
+   }
+
+   /* Done sending! Now, receive the reply: */
+
+   _mongoc_buffer_init (&buffer, NULL, 0, NULL, NULL);
+
+   if (!_mongoc_buffer_append_from_stream (&buffer,
+                                           server_monitor->stream,
+                                           4,
+                                           server_monitor->connect_timeout_ms,
+                                           error)) {
+      GOTO (fail);
```

Comment:
The `GOTO` macro is only applicable for the trace logging when the same function uses `ENTRY`, `EXIT`, and `RETURN`.

Reply:
I've used it here because _server_monitor_send_and_recv_opquery() also uses it (as far as I can tell also without the other macros!). I can rework that function as well as this one (the op_msg version) to use them, or just bare goto (a slight preference, but I can see the value of of trace logging also). Any preference?